### PR TITLE
feat: Add ability to update a dataset's metadata [CORE-2731]

### PR DESCRIPTION
**Pull request by joelvdavies**
Adds a command `dafni upload dataset-metadata` to upload new metadata to create a new metadata version on an existing dataset. Like #78 there is a --save option for saving existing metadata for modifying it manually.

Other changes:
- Renamed 'definition' to 'metadata' throughout the dataset commands as I thought it made it clearer

### CODE REVIEW

File: dafni_cli/datasets/dataset_upload.py [hunk 1]
```diff
@@ -169,3 +168,32 @@ def upload_dataset(
...
     click.echo(f"Metadata ID: {details['metadataId']}")
+
+
+def upload_dataset_metadata_version(
+    session: DAFNISession,
+    dataset_id: str,
+    version_id: str,
+    metadata: dict,
+) -> None:
+    """Function to upload a Dataset
```
Comment (sarahbyrnie) [inline dafni_cli/datasets/dataset_upload.py:179]: I don't think this is right

**Review by sarahbyrnie — COMMENTED** [review]
I think there's a test missing for the upload when the save option is chosen, other than that just one tiny comment but looks good

**Comment by joelvdavies** [comment]
> I think there's a test missing for the upload when the save option is chosen, other than that just one tiny comment but looks good

Ah yes I missed it out of the dataset-version as well. I have just added one for each.

File: test/commands/test_upload.py [hunk 2]
```diff
@@ -771,6 +834,56 @@ def test_upload_dataset_metadata_with_metadata_and_version_message(
...
         self.assertEqual(result.exit_code, 0)
 
+    def test_upload_metadata_cancel(
+        self,
+        mock_modify_dataset_metadata_for_upload,
+        mock_get_latest_dataset_metadata,
+        mock_upload_dataset_metadata_version,
+        mock_DAFNISession,
+    ):
+        """Tests that the 'upload dataset-version' command can be canceled"""
```
Comment (sarahbyrnie) [inline test/commands/test_upload.py:844]: upload dataset-metadata not version here

**Review by sarahbyrnie — APPROVED** [review]

**Pull request merged**